### PR TITLE
DataLoader with NamedTuple

### PR DESCRIPTION
**Pull request by cossio**
Just a couple of small changes, so that `DataLoader` can be created with a `NamedTuple` of tensors instead of `Tuple`. This way the tensors can be referred to by name. For example

```
train_loader = DataLoader((images = Xtrain, labels = Ytrain), batchsize=16)
batch = first(train_loader)
y = model(batch.images)
logitcrossentropy(y, batch.labels)
```

If we only use tuples, then in datasets with multiple tensors one has to be careful about the order in which the tensors are fed into the `DataLoader` constructor and be consistent with this elsewhere. With `NamedTuples` one just have to be consistent about the names used, which I think is a minor improvement.

CC @CarloLucibello 

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

I don't think this qualifies as an API change. It's just a minor feature addition. So final review probably not required.

- [x] Final review from `@MikeInnes` or `@dhairyagandhi96` (for API changes).

### CODE REVIEW

File: src/data/dataloader.jl [hunk 1]
```diff
@@ -16,8 +16,8 @@ end
 An object that iterates over mini-batches of `data`, each mini-batch containing `batchsize` observations
 (except possibly the last one). 
 
-Takes as input a data tensors or a tuple of one or more such tensors. 
-The last dimension in each tensor is considered to be the observation dimension. 
+Takes as input a data tensors or a tuple (or `NamedTuple`) of one or more such tensors.
```
Comment (CarloLucibello) [inline src/data/dataloader.jl:19]:
```suggestion
Takes as input a single data tensor, or a tuple (or a named tuple) of tensors.
```

File: src/data/dataloader.jl [hunk 2]
```diff
@@ -57,6 +57,13 @@ Usage example:
     # train for 10 epochs
     using IterTools: ncycle 
     Flux.train!(loss, ps, ncycle(train_loader, 10), opt)
+
+    # can use NamedTuple to name tensors
+    train_loader = DataLoader((images = Xtrain, labels = Ytrain), batchsize=2, shuffle=true)
```
Comment (CarloLucibello) [inline src/data/dataloader.jl:62]:
```suggestion
    train_loader = DataLoader((images=Xtrain, labels=Ytrain), batchsize=2, shuffle=true)
```

File: src/data/dataloader.jl [hunk 3]
```diff
@@ -88,19 +95,19 @@ end
 
 _nobs(data::AbstractArray) = size(data)[end]
 
-function _nobs(data::Tuple)
+function _nobs(data::Union{Tuple,NamedTuple})
```
Comment (CarloLucibello) [inline src/data/dataloader.jl:98]:
```suggestion
function _nobs(data::Union{Tuple, NamedTuple})
```

File: src/data/dataloader.jl [hunk 4]
```diff
@@ -88,19 +95,19 @@ end
...
     return n
 end
 
 function _getobs(data::AbstractArray{T,N}, i) where {T,N}
-    getindex(data, ntuple(i->Colon(), N-1)..., i)
+    getindex(data, ntuple(i->Colon(), Val(N-1))..., i)
 end
 
-_getobs(data::Tuple, i) = map(x -> _getobs(x, i), data)
+_getobs(data::Union{Tuple,NamedTuple}, i) = map(x -> _getobs(x, i), data)
```
Comment (CarloLucibello) [inline src/data/dataloader.jl:111]:
```suggestion
_getobs(data::Union{Tuple, NamedTuple}, i) = map(x -> _getobs(x, i), data)
```

File: test/data.jl [hunk 5]
```diff
@@ -38,6 +38,21 @@
     @test batches[3][1] == X[:,5:5]
     @test batches[3][2] == Y[5:5]
 
+    # test with NamedTuple
+    d = DataLoader((x = X, y = Y), batchsize=2)
```
Comment (CarloLucibello) [inline test/data.jl:42]:
```suggestion
    d = DataLoader((x=X, y=Y), batchsize=2)
```

**Comment by CarloLucibello** [comment]
I like this. Interaction with `train!` is  slightly problematic though, since fields will be splatted as positional arguments of the loss. Maybe we can consider this to be just ok

**Comment by dhairyagandhi96** [comment]
Rather than the Unions, would it make more sense have the NamedTuple constructors just forward to the regular ones.

Is there any other functional property that we are using here that warrants NamedTuple? Maybe it would be better to just expose these as kwargs in the first place, something like `data` and `labels` which could be their own tuples.

**Comment by cossio** [comment]
@CarloLucibello Hmm that's true. Maybe it's more consistent if we always call `loss(d)` instead of `loss(d...)`, and let the user handle `d` directly? But probably better to handle that elsewhere. In any case, using NamedTuples like this can be nicer for hand-written train loops.

@dhairyagandhi96 I don't understand? The idea is to be able to refer to the tensors by name, which can't be done if you convert the NamedTuple to a Tuple. Maybe I missunderstood your comment.

**Comment by dhairyagandhi96** [comment]
Looking from the code, the naming of said tensors is to allow users some convenience while sending the input, right? Or is the intention for the keys to be used inside the loss function?

**Comment by cossio** [comment]
> Looking from the code, the naming of said tensors is to allow users some convenience while sending the input, right? Or is the intention for the keys to be used inside the loss function?

I am using the keys inside the loss function (but I am also writing a training loop by hand). I think this could be a general use-case.

**Comment by cossio** [comment]
What do you think of https://github.com/FluxML/Flux.jl/issues/1227? I could try a fix here.

**Comment by CarloLucibello** [comment]
> What do you think of #1227? I could try a fix here.

that should be fixed, but better do it in a separate PR

**Comment by CarloLucibello** [comment]
In the end, DataLoader should end up supporting any type with some dataset-like interface. Changes here only involve overloading `_nobs` and `_getobs` for named tuples, twhich seems very lean and reasonable and goes in that direction, therefore I think we should merge this PR

File: src/data/dataloader.jl [hunk 6]
```diff
@@ -88,19 +95,16 @@ end
...
     end
     return n
 end
 
-function _getobs(data::AbstractArray{T,N}, i) where {T,N}
-    getindex(data, ntuple(i->Colon(), N-1)..., i)
-end
-
-_getobs(data::Tuple, i) = map(x -> _getobs(x, i), data)
+_getobs(data::AbstractArray, i) = data[ntuple(i -> Colon(), Val(ndims(data) - 1))..., i]
```
Comment (dhairyagandhi96) [inline src/data/dataloader.jl:107]: Use the `N` from the type and drop the `Val`?

File: src/data/dataloader.jl [hunk 7]
```diff
@@ -88,19 +95,16 @@ end
 
 _nobs(data::AbstractArray) = size(data)[end]
```
Comment (dhairyagandhi96) [inline src/data/dataloader.jl:96]: Use `size(data, N)` here

File: src/data/dataloader.jl (hunk 6 above)
Comment (cossio) [reply src/data/dataloader.jl:107]: Thanks for the suggestion, but why is that better?

**Comment by cossio** [comment]
Merge?

**Comment by CarloLucibello** [comment]
needs a rebase

**Comment by cossio** [comment]
rebased

**Comment by MikeInnes** [comment]
There are no 'minor' API changes that are allowed to go through without review; there are just API changes, and this is one of them. (The fact that you added documentation for it should be a giveaway).

I think this addition is fine though. It might be nice to generalise it (we could potentially reuse `Functor` here) but it's fine to make named tuples a special case wherever tuples are already, I think.

> Interaction with `train!` is slightly problematic though, since fields will be splatted as positional arguments of the loss.

It would be helpful to add a test to make sure the behaviour is right, but I don't think this is the case, eg:

```julia
julia> +((a=1,b=2)...)
3
```

**Comment by cossio** [comment]
@MikeInnes The problem with `train!` is that this line:

https://github.com/FluxML/Flux.jl/blob/254e4a70587efa08f720ab9d1fe9f9032ac61804/src/optimise/train.jl#L88

doesn't propagate the tensor names. So the user has to be careful to define the `loss` function to take arguments in the correct order.

**Comment by MikeInnes** [comment]
Ah, I misread Carlo's post. Yes, if we want to avoid that we'd have to do something significantly more complex and it's better to keep this simple.

One option is to write `train!(..., zip(DataLoader((a=a,b=b))`, in which case `loss` will be passed the named tuple directly.

**Comment by cossio** [comment]
In any case this `train!` issue can be dealt with in another PR / issue, right? I don't think more changes are needed here.

**Comment by MikeInnes** [comment]
Depends; this PR is something of a decision point, because if we wanted `train` to behave some other way with named tuples later, we'd have to break the behaviour added here. I think the behaviour here is probably only sensible one, though.

@CarloLucibello can decide if he's happy with the details but the current API change LGTM, anyway.

**Comment by CarloLucibello** [comment]
Once #1149 implementing https://github.com/FluxML/Flux.jl/pull/1149#issuecomment-620000631 gets merged, `train!` will pass the named tuple to the loss without splatting. Therefore inside train! we will have `loss(nt)`, which I think is better than `loss(nt...)` and essentially on par or even better than `loss(; nt...)`.  Users can also define define `loss(nt) = loss(; nt...)` if they really want to use keyword arguments. 

Let's merge this
bors r+

**Comment by cossio** [comment]
I do like `loss(; nt...)` to handle this ...

File: src/data/dataloader.jl (hunk 6 above)
Comment (dhairyagandhi96) [reply src/data/dataloader.jl:107]: It's largely for it to be cleaner, doing it like this doesn't seem to add any benefit and increases the complexity of the code while reading it
